cmd/kubeadm/app/images/interface_test.go

```diff
@@ -0,0 +1,274 @@
+func (f *imageExistsCmd) SetStdout(out io.Writer) {
+	f.out = out
+}
+func (f *imageExistsCmd) SetStderr(out io.Writer) {}
+func (f *imageExistsCmd) Stop()                   {}
+
+type imageExistsExecr struct {
+	cmd exec.Cmd
+}
+
+func (f *imageExistsExecr) Command(cmd string, args ...string) exec.Cmd { return f.cmd }
+func (f *imageExistsExecr) CommandContext(ctx context.Context, cmd string, args ...string) exec.Cmd {
+	return f.cmd
+}
+func (f *imageExistsExecr) LookPath(file string) (string, error) { return "something", nil }
+
+func TestImageExists(t *testing.T) {
+	testcases := []struct {
+		name        string
+		criSocket   string
+		image       string
+		expected    int
+		expectError bool
+		custombin   string
+	}{
+		{
+			name:        "using non-docker cri and image does not exist",
+			criSocket:   "/not/docker",
+			image:       "image:does-not-exist",
+			expected:    0,
+			expectError: false,
+		},
+		{
+			name:        "using non-docker cri and image does exist",
+			criSocket:   "/not/docker",
+			image:       "image:exists",
+			expected:    1,
+			expectError: false,
+		},
+		{
+			name:        "using docker and image exists",
+			criSocket:   kubeadmapiv1alpha2.DefaultCRISocket,
+			image:       "image:exists",
+			expected:    1,
+			expectError: false,
+		},
+		{
+			name:        "using docker and image does not exist",
+			criSocket:   kubeadmapiv1alpha2.DefaultCRISocket,
+			image:       "image:does-not-exist",
+			expected:    0,
+			expectError: false,
+		},
+		{
+			name:        "use docker but docker errors",
+			criSocket:   kubeadmapiv1alpha2.DefaultCRISocket,
+			image:       "bogusimage:latest",
+			expected:    -1,
+			expectError: true,
+		},
+		{
+			name:        "can't find crictl binary but we want to use crictl",
+			criSocket:   "bogus",
+			image:       "whatever:1.0",
+			expected:    -1,
+			expectError: true,
+			custombin:   "not-a-real-binary",
+		},
+	}
+
+	for _, tc := range testcases {
+		t.Run(tc.name, func(t *testing.T) {
+			var ie images.Existence
+			cri := "crictl"
+			if tc.criSocket == kubeadmapiv1alpha2.DefaultCRISocket {
+				cri = "docker"
+			}
+
+			if tc.custombin != "" {
+				cri = tc.custombin
+			}
+
+			iee := &imageExistsExecr{
+				cmd: &imageExistsCmd{
+					image: tc.image,
+					cri:   cri,
+				},
+			}
+			// error is checked in TestImagePuller
+			ie, _ = images.NewCRInterfacer(iee, tc.criSocket)
+
+			found, err := ie.Exists(tc.image)
+			if found != tc.expected {
+				t.Fatalf("expected %v but found %v: %v", tc.expected, found, err)
+			}
+			if tc.expectError {
+				if err == nil {
+					t.Fatal("expected an error but got nil")
+				}
+			}
+		})
+	}
+}
+
+func TestImagePuller(t *testing.T) {
+	testcases := []struct {
+		name          string
+		criSocket     string
+		cmd           exec.Cmd
+		findCrictl    bool
+		expected      string
+		errorExpected bool
+	}{
+		{
+			name:      "New succeeds even if crictl is not in path",
```

Comment:
These seems like unit tests for `NewCRInterfacer` rather than the pull functionality. Maybe you can keep this, make it focus only on `NewCRInterfacer`, and then create one more unit test to specifically test the pull functionality?